### PR TITLE
 feat(codecs): Implement JSON codec

### PR DESCRIPTION
**Pull request by pablosichert**
Depends on https://github.com/timberio/vector/pull/7482.

Part of https://github.com/timberio/vector/issues/7465.

### CODE REVIEW

**Review by spencergilbert — APPROVED** [review]

File: src/codecs/json.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        let mut output = transform.transform(input);
+        let transformed = output.next().await.unwrap();
+
+        assert_eq!(
+            transformed
+                .as_log()
+                .get(log_schema().message_key())
+                .unwrap()
+                .clone(),
+            r#"{"message":{"foo":"bar"}}"#.to_owned().into()
```
Comment (pablosichert) [inline src/codecs/json.rs:172]: I'm a bit unsure about exposing the "message" field when not specifying the source field. However, there doesn't seem to be an unambiguous way to differentiate between an object on the root that incidentally has a message key, or the "main" message key from the log schema.

File: src/codecs/json.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                }
+
+                let target_field = options
+                    .target_field
+                    .as_ref()
+                    .map(AsRef::as_ref)
+                    .unwrap_or_else(|| log_schema().message_key());
+
+                let contains_target = log.contains(&target_field);
+                if !contains_target || options.overwrite_target {
```
Comment (bruceg) [inline src/codecs/json.rs:88]: Can these conditions be reversed so that the `log.contains` is only evaluated if `options.overwrite_target` is false?

**Review by bruceg — COMMENTED** [review]
This looks reasonable and helps elaborate on what's going on in #7482. I will note again this is undocumented though.

File: src/codecs/json.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    config::log_schema,
+    event::{Event, Value},
+    transform::{FunctionTransform, Transform},
+};
+
+#[derive(Debug, Clone, Serialize, Deserialize, Derivative)]
+#[serde(deny_unknown_fields, default)]
+#[derivative(Default)]
+pub struct JsonCodec {
+    mode: Option<Mode>,
```
Comment (lukesteensen) [inline src/codecs/json.rs:17]: This doesn't seem to be used for anything.

File: src/codecs/json.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+                let target_field = options
+                    .target_field
+                    .as_ref()
+                    .map(AsRef::as_ref)
+                    .unwrap_or_else(|| log_schema().message_key());
+
+                let contains_target = log.contains(&target_field);
+                if !contains_target || options.overwrite_target {
+                    log.insert(&target_field, serialized);
```
Comment (lukesteensen) [inline src/codecs/json.rs:89]: This is a good example of why I find the whole `Transform`-based model confusing. We're serializing an event into a JSON string and then reinserting the serialized data back into itself? So there's no way to distinguish between a normal event and one that's been serialized, and sinks will have to know to only forward the one serialized field?

File: src/codecs/json.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+                let json = match &options.field {
+                    Some(field) => match log.get(field) {
+                        Some(value) => value,
+                        None => &Value::Null,
+                    },
+                    None => log.get_root(),
+                };
+
+                let serialized = match serde_json::to_string(json) {
```
Comment (lukesteensen) [inline src/codecs/json.rs:65]: Unconditionally serializing into a string before even hitting the sink is going to negatively impact performance. In a number of cases I believe we're currently serializing out directly to a socket/file/etc buffer and avoiding these intermediate allocations.

**Review by lukesteensen — CHANGES_REQUESTED** [review]
See my review on #7352 for a broader discussion. Also, have you tried integrating this with a sink like ES or similar? It's interesting to see what a concrete implementation looks like, but not nearly as helpful as seeing whether or not the API integrates smoothly with the intended consumers.

File: src/codecs/json.rs (hunk 3 above)
Comment (pablosichert) [reply src/codecs/json.rs:17]: Oops, yes I accidentally left that in from the previous (admittedly weird) feature: https://github.com/timberio/vector/pull/7803/commits/f4e290f23e2d5f2aaacda3e39d1640d44f114b45#diff-61a423b9da07f268a42699f3bf330341a5a596b16439332c9f98b704c776ae73R104-R109

**Comment by pablosichert** [comment]
Closing this with https://github.com/timberio/vector/pull/7482.

**Pull request closed without merging**